src/cpp/fastdds/builtin/type_lookup_service/TypeLookupManager.cpp

```diff
@@ -311,79 +311,70 @@ SampleIdentity TypeLookupManager::get_types(
 }
 
 ReturnCode_t TypeLookupManager::async_get_type(
-        xtypes::TypeInformation type_inf,
+        xtypes::TypeInformation type_information,
         fastrtps::rtps::GuidPrefix_t type_server,
         AsyncGetTypeCallback& callback)
 {
-    ReturnCode_t solve_ret = solve_type(type_inf.complete().typeid_with_size());
+    ReturnCode_t solve_ret = check_type_identifier_received(type_information.complete().typeid_with_size());
     if (RETCODE_OK == solve_ret)
     {
-        // The type is already known, invoke the callback and return success
+        // The type is already known, invoke the callback
         callback();
     }
     else if (RETCODE_NO_DATA == solve_ret)
     {
-        // The type is being solved, add callback to notify
-        add_async_get_type_callback(type_inf.complete().typeid_with_size(), type_server, callback);
+        // The type needs to be queried, add callback to notify once the remote type has been discovered.
+        add_async_get_type_callback(type_information.complete().typeid_with_size(), type_server, callback);
     }
+
     return solve_ret;
 }
 
-ReturnCode_t TypeLookupManager::solve_type(
-        xtypes::TypeIdentfierWithSize type_id)
+ReturnCode_t TypeLookupManager::check_type_identifier_received(
+        xtypes::TypeIdentfierWithSize type_identifier_with_size)
 {
-    xtypes::TypeIdentifierSeq uknown_type;
-
     // Check if the type is known
     if (fastrtps::rtps::RTPSDomainImpl::get_instance()->type_object_registry_observer().
-                    is_type_identifier_known(type_id.type_id()))
+                    is_type_identifier_known(type_identifier_with_size))
     {
         return RETCODE_OK;
     }
-    else
+
+    // Check if the TypeInformation was received before
+    if (async_get_type_callbacks_.find(type_identifier_with_size) != async_get_type_callbacks_.end())
```

Comment:
In this case you should register this new callback in the async_get_type_callbacks_ collection, within the same TypeIdentfierWithSize key, adding a new element to the AsyncGetTypeCallback vector. Consequently, to avoid multiple searchs, the iterator can be saved in a variable, check if it is the end iterator or not, and if it is not, accesing the iterator element and adding the information to the corresponding vector.

Reply:
Changed in [cfa3f78](https://github.com/eProsima/Fast-DDS/pull/4177/commits/cfa3f781c09d678528b673acc9feef170c546bc0)